modules/core/src/main/scala/gem/ProgramId.scala

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2016-2017 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package gem
+
+import java.time._
+import java.time.format.DateTimeFormatter
+import gem.enum.{ Site, ProgramType, DailyProgramType }
+import scalaz._, scalaz.Scalaz.{ char => _, _ }
+
+/**
+ * A science program id, which has three constructors. `Science` for standard programs; `Daily` for
+ * standard daily engineering and calibration programs; and `Nonstandard` for all others.
+ */
+sealed abstract class ProgramId(
+  val siteOption:        Option[Site],
+  val semesterOption:    Option[Semester],
+  val programTypeOption: Option[ProgramType]
+) extends Product with Serializable {
+
+  /**
+   * Format a canonical string for this `ProgramId`. This is for human use, and can be round-
+   * tripped through `ProgramId.fromString`.
+   */
+  def format: String
+
+}
+
+object ProgramId {
+
+  /** A standard science program id with a site, semester, program type, and positive index. */
+  sealed abstract case class Science private (
+    site:        Site,
+    semester:    Semester,
+    programType: ProgramType,
+    index:       Int
+  ) extends ProgramId(
+    Some(site),
+    Some(semester),
+    Some(programType)
+  ) {
+
+    // A copy method isn't generated because this class is abstract, so we do so here. It can't
+    // support `index` because this could break the invariant that it must be positive.
+    @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
+    def copy(
+      site:        Site        = this.site,
+      semester:    Semester    = this.semester,
+      programType: ProgramType = this.programType
+    ): Science =
+      Science.unsafeApply(site, semester, programType, index)
+
+    override def format =
+      s"${site.shortName}-${semester.format}-${programType.shortName}-$index"
+
+  }
+  object Science {
+
+    /** Construct a `Science` program id, if `index` is positive. */
+    def apply(site: Site, semester: Semester, programType: ProgramType, index: Int): Option[Science] =
```

Comment:
Yeah it would be useful. Perhaps something like

```
case class UnsignedNum[A](val a: A)(implicit n: Numeric[A])
```

You'd still get an option on the `apply` though